Flaky test: 10/10k: Test/HTTPHeaderFrameErrorHandlingNormalTrailer & Test/HTTPHeaderFrameErrorHandlingMoreThanTwoHeaders

https://github.com/grpc/grpc-go/runs/4232539138?check_suite_focus=true


```
--- FAIL: Test (45.27s)
    --- FAIL: Test/HTTPHeaderFrameErrorHandlingNormalTrailer (0.00s)
        tlogger.go:116: INFO clientconn.go:175 [core] Channel Created  (t=+344.504µs)
        tlogger.go:116: INFO clientconn.go:1623 [core] original dial target is: "127.0.0.1:35051"  (t=+365.604µs)
        tlogger.go:116: INFO clientconn.go:1628 [core] dial target "127.0.0.1:35051" parse failed: parse "127.0.0.1:35051": first path segment in URL cannot contain colon  (t=+402.405µs)
        tlogger.go:116: INFO clientconn.go:1643 [core] fallback to scheme "passthrough"  (t=+418.005µs)
        tlogger.go:116: INFO clientconn.go:1651 [core] parsed dial target is: {Scheme:passthrough Authority: Endpoint:127.0.0.1:35051 URL:{Scheme:passthrough Opaque: User: Host: Path:/127.0.0.1:35051 RawPath: ForceQuery:false RawQuery: Fragment: RawFragment:}}  (t=+438.205µs)
        tlogger.go:116: INFO clientconn.go:262 [core] Channel authority set to "127.0.0.1:35051"  (t=+451.305µs)
        tlogger.go:116: INFO resolver_conn_wrapper.go:100 [core] ccResolverWrapper: sending update to cc: {[{127.0.0.1:35051  <nil> <nil> 0 <nil>}] <nil> <nil>}  (t=+467.705µs)
        tlogger.go:116: INFO resolver_conn_wrapper.go:183 [core] Resolver state updated: {Addresses:[{Addr:127.0.0.1:35051 ServerName: Attributes:<nil> BalancerAttributes:<nil> Type:0 Metadata:<nil>}] ServiceConfig:<nil> Attributes:<nil>} (resolver returned new addresses)  (t=+483.906µs)
        tlogger.go:116: INFO clientconn.go:712 [core] ClientConn switching balancer to "pick_first"  (t=+495.706µs)
        tlogger.go:116: INFO clientconn.go:732 [core] Channel switches to new LB policy "pick_first"  (t=+507.806µs)
        tlogger.go:116: INFO clientconn.go:773 [core] Subchannel Created  (t=+534.706µs)
        tlogger.go:116: INFO clientconn.go:773 [core] Subchannel(id:1231) created  (t=+563.906µs)
        tlogger.go:116: INFO picker_wrapper.go:161 [core] blockingPicker: the picked transport is not ready, loop back to repick  (t=+605.507µs)
        tlogger.go:116: INFO clientconn.go:1143 [core] Subchannel Connectivity change to CONNECTING  (t=+654.808µs)
        tlogger.go:116: INFO clientconn.go:1254 [core] Subchannel picks a new address "127.0.0.1:35051" to connect  (t=+682.608µs)
        tlogger.go:116: INFO clientconn.go:417 [core] Channel Connectivity change to CONNECTING  (t=+810.509µs)
        tlogger.go:116: INFO clientconn.go:1143 [core] Subchannel Connectivity change to READY  (t=+988.411µs)
        tlogger.go:116: INFO clientconn.go:417 [core] Channel Connectivity change to READY  (t=+1.021012ms)
        tlogger.go:116: INFO clientconn.go:417 [core] Channel Connectivity change to SHUTDOWN  (t=+1.214114ms)
        tlogger.go:116: INFO clientconn.go:1143 [core] Subchannel Connectivity change to SHUTDOWN  (t=+1.246914ms)
        tlogger.go:116: INFO clientconn.go:1501 [core] Subchannel Deleted  (t=+1.260315ms)
        tlogger.go:116: INFO clientconn.go:1501 [core] Subchanel(id:1231) deleted  (t=+1.269615ms)
        tlogger.go:116: INFO clientconn.go:1099 [core] Channel Deleted  (t=+1.279815ms)
        tlogger.go:116: INFO types.go:68 [channelz] attempt to delete child with id 1231 from a parent (id=1230) that doesn't currently exist  (t=+1.428417ms)
        tlogger.go:116: INFO types.go:68 [channelz] attempt to delete child with id 1231 from a parent (id=1230) that doesn't currently exist  (t=+1.487917ms)
        tlogger.go:116: INFO clientconn.go:175 [core] Channel Created  (t=+1.513418ms)
        tlogger.go:116: INFO clientconn.go:1623 [core] original dial target is: "127.0.0.1:38583"  (t=+1.527718ms)
        tlogger.go:116: INFO clientconn.go:1628 [core] dial target "127.0.0.1:38583" parse failed: parse "127.0.0.1:38583": first path segment in URL cannot contain colon  (t=+1.546718ms)
        tlogger.go:116: INFO clientconn.go:1643 [core] fallback to scheme "passthrough"  (t=+1.578018ms)
        tlogger.go:116: INFO clientconn.go:1651 [core] parsed dial target is: {Scheme:passthrough Authority: Endpoint:127.0.0.1:38583 URL:{Scheme:passthrough Opaque: User: Host: Path:/127.0.0.1:38583 RawPath: ForceQuery:false RawQuery: Fragment: RawFragment:}}  (t=+1.600019ms)
        tlogger.go:116: INFO clientconn.go:262 [core] Channel authority set to "127.0.0.1:38583"  (t=+1.613319ms)
        tlogger.go:116: INFO resolver_conn_wrapper.go:100 [core] ccResolverWrapper: sending update to cc: {[{127.0.0.1:38583  <nil> <nil> 0 <nil>}] <nil> <nil>}  (t=+1.629119ms)
        tlogger.go:116: INFO resolver_conn_wrapper.go:183 [core] Resolver state updated: {Addresses:[{Addr:127.0.0.1:38583 ServerName: Attributes:<nil> BalancerAttributes:<nil> Type:0 Metadata:<nil>}] ServiceConfig:<nil> Attributes:<nil>} (resolver returned new addresses)  (t=+1.644719ms)
        tlogger.go:116: INFO clientconn.go:712 [core] ClientConn switching balancer to "pick_first"  (t=+1.680419ms)
        tlogger.go:116: INFO clientconn.go:732 [core] Channel switches to new LB policy "pick_first"  (t=+1.69482ms)
        tlogger.go:116: INFO clientconn.go:773 [core] Subchannel Created  (t=+1.72962ms)
        tlogger.go:116: INFO clientconn.go:773 [core] Subchannel(id:1234) created  (t=+1.75762ms)
        tlogger.go:116: INFO picker_wrapper.go:161 [core] blockingPicker: the picked transport is not ready, loop back to repick  (t=+1.784921ms)
        tlogger.go:116: INFO clientconn.go:1143 [core] Subchannel Connectivity change to CONNECTING  (t=+1.805221ms)
        tlogger.go:116: INFO clientconn.go:1254 [core] Subchannel picks a new address "127.0.0.1:38583" to connect  (t=+1.825621ms)
        tlogger.go:116: INFO clientconn.go:417 [core] Channel Connectivity change to CONNECTING  (t=+1.898922ms)
        tlogger.go:116: INFO clientconn.go:1143 [core] Subchannel Connectivity change to READY  (t=+2.139025ms)
        tlogger.go:116: INFO clientconn.go:417 [core] Channel Connectivity change to READY  (t=+2.164725ms)
        tlogger.go:116: INFO clientconn.go:417 [core] Channel Connectivity change to SHUTDOWN  (t=+2.321927ms)
        tlogger.go:116: INFO clientconn.go:1143 [core] Subchannel Connectivity change to SHUTDOWN  (t=+2.338427ms)
        tlogger.go:116: INFO clientconn.go:1501 [core] Subchannel Deleted  (t=+2.350927ms)
        tlogger.go:116: INFO clientconn.go:1501 [core] Subchanel(id:1234) deleted  (t=+2.360727ms)
        tlogger.go:116: INFO clientconn.go:1099 [core] Channel Deleted  (t=+2.370727ms)
        tlogger.go:116: INFO clientconn.go:175 [core] Channel Created  (t=+2.541729ms)
        tlogger.go:116: INFO types.go:68 [channelz] attempt to delete child with id 1234 from a parent (id=1233) that doesn't currently exist  (t=+2.56013ms)
        tlogger.go:116: INFO types.go:68 [channelz] attempt to delete child with id 1234 from a parent (id=1233) that doesn't currently exist  (t=+2.59283ms)
        tlogger.go:116: INFO clientconn.go:1623 [core] original dial target is: "127.0.0.1:42691"  (t=+2.62533ms)
        tlogger.go:116: INFO clientconn.go:1628 [core] dial target "127.0.0.1:42691" parse failed: parse "127.0.0.1:42691": first path segment in URL cannot contain colon  (t=+2.646431ms)
        tlogger.go:116: INFO clientconn.go:1643 [core] fallback to scheme "passthrough"  (t=+2.680731ms)
        tlogger.go:116: INFO clientconn.go:1651 [core] parsed dial target is: {Scheme:passthrough Authority: Endpoint:127.0.0.1:42691 URL:{Scheme:passthrough Opaque: User: Host: Path:/127.0.0.1:42691 RawPath: ForceQuery:false RawQuery: Fragment: RawFragment:}}  (t=+2.703631ms)
        end2end_test.go:7418: Error at server-side while reading frame. Err: read tcp 127.0.0.1:38583->127.0.0.1:36996: read: connection reset by peer
        tlogger.go:116: INFO clientconn.go:262 [core] Channel authority set to "127.0.0.1:42691"  (t=+2.716931ms)
        tlogger.go:116: INFO resolver_conn_wrapper.go:100 [core] ccResolverWrapper: sending update to cc: {[{127.0.0.1:42691  <nil> <nil> 0 <nil>}] <nil> <nil>}  (t=+2.758132ms)
        tlogger.go:116: INFO resolver_conn_wrapper.go:183 [core] Resolver state updated: {Addresses:[{Addr:127.0.0.1:42691 ServerName: Attributes:<nil> BalancerAttributes:<nil> Type:0 Metadata:<nil>}] ServiceConfig:<nil> Attributes:<nil>} (resolver returned new addresses)  (t=+2.775732ms)
        tlogger.go:116: INFO clientconn.go:712 [core] ClientConn switching balancer to "pick_first"  (t=+2.803832ms)
        tlogger.go:116: INFO clientconn.go:732 [core] Channel switches to new LB policy "pick_first"  (t=+2.816733ms)
        tlogger.go:116: INFO clientconn.go:773 [core] Subchannel Created  (t=+2.834333ms)
        tlogger.go:116: INFO clientconn.go:773 [core] Subchannel(id:1237) created  (t=+2.844733ms)
        tlogger.go:116: INFO picker_wrapper.go:161 [core] blockingPicker: the picked transport is not ready, loop back to repick  (t=+2.867533ms)
        tlogger.go:116: INFO clientconn.go:1143 [core] Subchannel Connectivity change to CONNECTING  (t=+2.886033ms)
        tlogger.go:116: INFO clientconn.go:1254 [core] Subchannel picks a new address "127.0.0.1:42691" to connect  (t=+2.897134ms)
        tlogger.go:116: INFO clientconn.go:417 [core] Channel Connectivity change to CONNECTING  (t=+3.029735ms)
        tlogger.go:116: INFO clientconn.go:1143 [core] Subchannel Connectivity change to READY  (t=+3.125536ms)
        tlogger.go:116: INFO clientconn.go:417 [core] Channel Connectivity change to READY  (t=+3.156337ms)
        tlogger.go:116: INFO clientconn.go:417 [core] Channel Connectivity change to SHUTDOWN  (t=+3.328339ms)
        tlogger.go:116: INFO clientconn.go:1143 [core] Subchannel Connectivity change to SHUTDOWN  (t=+3.346039ms)
        tlogger.go:116: INFO clientconn.go:1501 [core] Subchannel Deleted  (t=+3.358939ms)
        tlogger.go:116: INFO clientconn.go:1501 [core] Subchanel(id:1237) deleted  (t=+3.369139ms)
        tlogger.go:116: INFO clientconn.go:1099 [core] Channel Deleted  (t=+3.394439ms)
        tlogger.go:116: INFO types.go:68 [channelz] attempt to delete child with id 1237 from a parent (id=1236) that doesn't currently exist  (t=+3.49324ms)
        tlogger.go:116: INFO types.go:68 [channelz] attempt to delete child with id 1237 from a parent (id=1236) that doesn't currently exist  (t=+3.518541ms)
FAIL
FAIL	google.golang.org/grpc/test	107.008s
```

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -902,7 +902,7 @@ const possibleConnResetMsg = "connection reset by peer"
 const possibleEOFMsg = "error reading from server: EOF"
 
 func errMsgContainsPossibleStopMessages(err error) bool {
-	return !strings.Contains(err.Error(), possibleConnResetMsg) && !strings.Contains(err.Error(), possibleEOFMsg)
+	return strings.Contains(err.Error(), possibleConnResetMsg) || strings.Contains(err.Error(), possibleEOFMsg) || err == io.EOF
 }
 
 func (l *lazyConn) Write(b []byte) (int, error) {
@@ -1046,7 +1046,7 @@ func (s) TestDetailedConnectionCloseErrorPropagatesToRpcError(t *testing.T) {
 	// The precise behavior of this test is subject to raceyness around the timing of when TCP packets
 	// are sent from client to server, and when we tell the server to stop, so we need to account for both
 	// of these possible error messages
-	if _, err := stream.Recv(); err == nil || errMsgContainsPossibleStopMessages(err) {
+	if _, err := stream.Recv(); err == nil || !errMsgContainsPossibleStopMessages(err) {
 		t.Fatalf("%v.Recv() = _, %v, want _, rpc error containing substring: %q OR %q", stream, err, possibleConnResetMsg, possibleEOFMsg)
 	}
 	close(rpcDoneOnClient)
@@ -6953,7 +6953,7 @@ func (s *httpServer) start(t *testing.T, lis net.Listener) {
 			for {
 				frame, err := framer.ReadFrame()
 				if err != nil {
-					if errMsgContainsPossibleStopMessages(err) {
+					if !errMsgContainsPossibleStopMessages(err) {
 						t.Errorf("Error at server-side while reading frame. got: %q, want: rpc error containing substring %q OR %q", err, possibleConnResetMsg, possibleEOFMsg)
 					}
 					return

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -901,11 +901,12 @@ const possibleEOFMsg = "error reading from server: EOF"
 //     the kernel can send an RST back to the client (also see
 //     https://stackoverflow.com/questions/33053507/econnreset-in-send-linux-c).
 //     Note that while this condition is expected to be rare due to the
-//     rpcStartedOnServer synchronization, in theory it should be possible,
+//     test httpServer start synchronization, in theory it should be possible,
 //     e.g. if the client sends a BDP ping at the right time.
 //  2. If, for example, the call to ss.S.Stop() happens after the RPC headers
 //     have been received at the server, then the TCP connection can shutdown
 //     gracefully when the server's socket closes.
+//  3. If there is an actual io.EOF received because the client stopped the stream.
 func isConnClosedErr(err error) bool {
 	errContainsConnResetMsg := strings.Contains(err.Error(), possibleConnResetMsg)
 	errContainsEOFMsg := strings.Contains(err.Error(), possibleEOFMsg)
@@ -1054,7 +1055,7 @@ func (s) TestDetailedConnectionCloseErrorPropagatesToRpcError(t *testing.T) {
 	// The precise behavior of this test is subject to raceyness around the timing
 	// of when TCP packets are sent from client to server, and when we tell the
 	// server to stop, so we need to account for both possible error messages.
-	if _, err := stream.Recv(); err == nil || !isConnClosedErr(err) {
+	if _, err := stream.Recv(); !isConnClosedErr(err) {
 		t.Fatalf("%v.Recv() = _, %v, want _, rpc error containing substring: %q OR %q", stream, err, possibleConnResetMsg, possibleEOFMsg)
 	}
 	close(rpcDoneOnClient)
@@ -7034,7 +7035,7 @@ func (s *httpServer) start(t *testing.T, lis net.Listener) {
 func doHTTPHeaderTest(t *testing.T, errCode codes.Code, headerFields ...[]string) error {
 	lis, err := net.Listen("tcp", "localhost:0")
 	if err != nil {
-		return fmt.Errorf("failed to listen. Err: %v", err)
+		return fmt.Errorf("listening on %q: %v", "localhost:0", err)
 	}
 	defer lis.Close()
 	server := &httpServer{
@@ -7043,7 +7044,7 @@ func doHTTPHeaderTest(t *testing.T, errCode codes.Code, headerFields ...[]string
 	server.start(t, lis)
 	cc, err := grpc.Dial(lis.Addr().String(), grpc.WithTransportCredentials(insecure.NewCredentials()))
 	if err != nil {
-		return fmt.Errorf("failed to dial due to err: %v", err)
+		return fmt.Errorf("dial(%q): %v", lis.Addr().String(), err)
 	}
 	defer cc.Close()
 	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
@@ -7051,10 +7052,10 @@ func doHTTPHeaderTest(t *testing.T, errCode codes.Code, headerFields ...[]string
 	client := testpb.NewTestServiceClient(cc)
 	stream, err := client.FullDuplexCall(ctx)
 	if err != nil {
-		return fmt.Errorf("error creating stream due to err: %v", err)
+		return fmt.Errorf("creating FullDuplex stream: %v", err)
 	}
 	if _, err := stream.Recv(); err == nil || status.Code(err) != errCode {
-		return fmt.Errorf("stream.Recv() = _, %v, want error code: %v", err, errCode)
+		return fmt.Errorf("stream.Recv() = %v, want error code: %v", err, errCode)
 	}
 	return nil
 }

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -888,6 +888,23 @@ type lazyConn struct {
 	beLazy int32
 }
 
+// errMsgContainsPossibleStopMessages addresses around the timing of when TCP packets
+// are sent from client to server, and when we tell the server to stop, so we need to account for both
+// of these possible error messages:
+//  1. If the call to ss.S.Stop() causes the server's sockets to close while there's still in-fight
+//     data from the client on the TCP connection, then the kernel can send an RST back to the client (also
+//     see https://stackoverflow.com/questions/33053507/econnreset-in-send-linux-c). Note that while this
+//     condition is expected to be rare due to the rpcStartedOnServer synchronization, in theory it should
+//     be possible, e.g. if the client sends a BDP ping at the right time.
+//  2. If, for example, the call to ss.S.Stop() happens after the RPC headers have been received at the
+//     server, then the TCP connection can shutdown gracefully when the server's socket closes. raceyness around checks if the error message when checking reading frames
+const possibleConnResetMsg = "connection reset by peer"
+const possibleEOFMsg = "error reading from server: EOF"
+
+func errMsgContainsPossibleStopMessages(err error) bool {
+	return !strings.Contains(err.Error(), possibleConnResetMsg) && !strings.Contains(err.Error(), possibleEOFMsg)
+}
+
 func (l *lazyConn) Write(b []byte) (int, error) {
 	if atomic.LoadInt32(&(l.beLazy)) == 1 {
 		time.Sleep(time.Second)
@@ -1013,18 +1030,7 @@ func (s) TestDetailedConnectionCloseErrorPropagatesToRpcError(t *testing.T) {
 
 	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
 	defer cancel()
-	// The precise behavior of this test is subject to raceyness around the timing of when TCP packets
-	// are sent from client to server, and when we tell the server to stop, so we need to account for both
-	// of these possible error messages:
-	// 1) If the call to ss.S.Stop() causes the server's sockets to close while there's still in-fight
-	//    data from the client on the TCP connection, then the kernel can send an RST back to the client (also
-	//    see https://stackoverflow.com/questions/33053507/econnreset-in-send-linux-c). Note that while this
-	//    condition is expected to be rare due to the rpcStartedOnServer synchronization, in theory it should
-	//    be possible, e.g. if the client sends a BDP ping at the right time.
-	// 2) If, for example, the call to ss.S.Stop() happens after the RPC headers have been received at the
-	//    server, then the TCP connection can shutdown gracefully when the server's socket closes.
-	const possibleConnResetMsg = "connection reset by peer"
-	const possibleEOFMsg = "error reading from server: EOF"
+
 	// Start an RPC. Then, while the RPC is still being accepted or handled at the server, abruptly
 	// stop the server, killing the connection. The RPC error message should include details about the specific
 	// connection error that was encountered.
@@ -1037,7 +1043,10 @@ func (s) TestDetailedConnectionCloseErrorPropagatesToRpcError(t *testing.T) {
 	// the RPC has been started on it.
 	<-rpcStartedOnServer
 	ss.S.Stop()
-	if _, err := stream.Recv(); err == nil || (!strings.Contains(err.Error(), possibleConnResetMsg) && !strings.Contains(err.Error(), possibleEOFMsg)) {
+	// The precise behavior of this test is subject to raceyness around the timing of when TCP packets
+	// are sent from client to server, and when we tell the server to stop, so we need to account for both
+	// of these possible error messages
+	if _, err := stream.Recv(); err == nil || errMsgContainsPossibleStopMessages(err) {
 		t.Fatalf("%v.Recv() = _, %v, want _, rpc error containing substring: %q OR %q", stream, err, possibleConnResetMsg, possibleEOFMsg)
 	}
 	close(rpcDoneOnClient)
@@ -6815,12 +6824,14 @@ func (s) TestHTTPHeaderFrameErrorHandlingInitialHeader(t *testing.T) {
 
 // Testing non-Trailers-only Trailers (delivered in second HEADERS frame)
 func (s) TestHTTPHeaderFrameErrorHandlingNormalTrailer(t *testing.T) {
-	for _, test := range []struct {
+	tests := []struct {
+		name           string
 		responseHeader []string
 		trailer        []string
 		errCode        codes.Code
 	}{
 		{
+			name: "trailer missing grpc-status",
 			responseHeader: []string{
 				":status", "200",
 				"content-type", "application/grpc",
@@ -6832,6 +6843,7 @@ func (s) TestHTTPHeaderFrameErrorHandlingNormalTrailer(t *testing.T) {
 			errCode: codes.Unavailable,
 		},
 		{
+			name: "malformed grpc-status-details-bin field with status 404",
 			responseHeader: []string{
 				":status", "404",
 				"content-type", "application/grpc",
@@ -6844,6 +6856,7 @@ func (s) TestHTTPHeaderFrameErrorHandlingNormalTrailer(t *testing.T) {
 			errCode: codes.Unimplemented,
 		},
 		{
+			name: "malformed grpc-status-details-bin field with status 200",
 			responseHeader: []string{
 				":status", "200",
 				"content-type", "application/grpc",
@@ -6855,8 +6868,12 @@ func (s) TestHTTPHeaderFrameErrorHandlingNormalTrailer(t *testing.T) {
 			},
 			errCode: codes.Internal,
 		},
-	} {
-		doHTTPHeaderTest(t, test.errCode, test.responseHeader, test.trailer)
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			doHTTPHeaderTest(t, test.errCode, test.responseHeader, test.trailer)
+		})
+
 	}
 }
 
@@ -6930,14 +6947,14 @@ func (s *httpServer) start(t *testing.T, lis net.Listener) {
 		}
 		writer.Flush() // necessary since client is expecting preface before declaring connection fully setup.
 		var sid uint32
-		// Loop until conn is closed and framer returns io.EOF
+		// Loop until conn is closed and framer returns possible Stop error messages
 		for requestNum := 0; ; requestNum = (requestNum + 1) % len(s.responses) {
 			// Read frames until a header is received.
 			for {
 				frame, err := framer.ReadFrame()
 				if err != nil {
-					if err != io.EOF {
-						t.Errorf("Error at server-side while reading frame. Err: %v", err)
+					if errMsgContainsPossibleStopMessages(err) {
+						t.Errorf("Error at server-side while reading frame. got: %q, want: rpc error containing substring %q OR %q", err, possibleConnResetMsg, possibleEOFMsg)
 					}
 					return
 				}

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -1055,7 +1055,7 @@ func (s) TestDetailedConnectionCloseErrorPropagatesToRpcError(t *testing.T) {
 	// The precise behavior of this test is subject to raceyness around the timing
 	// of when TCP packets are sent from client to server, and when we tell the
 	// server to stop, so we need to account for both possible error messages.
-	if _, err := stream.Recv(); !isConnClosedErr(err) {
+	if _, err := stream.Recv(); err == io.EOF || !isConnClosedErr(err) {
 		t.Fatalf("%v.Recv() = _, %v, want _, rpc error containing substring: %q OR %q", stream, err, possibleConnResetMsg, possibleEOFMsg)
 	}
 	close(rpcDoneOnClient)

--- a/test/end2end_test.go
+++ b/test/end2end_test.go
@@ -888,21 +888,29 @@ type lazyConn struct {
 	beLazy int32
 }
 
-// errMsgContainsPossibleStopMessages addresses around the timing of when TCP packets
-// are sent from client to server, and when we tell the server to stop, so we need to account for both
-// of these possible error messages:
-//  1. If the call to ss.S.Stop() causes the server's sockets to close while there's still in-fight
-//     data from the client on the TCP connection, then the kernel can send an RST back to the client (also
-//     see https://stackoverflow.com/questions/33053507/econnreset-in-send-linux-c). Note that while this
-//     condition is expected to be rare due to the rpcStartedOnServer synchronization, in theory it should
-//     be possible, e.g. if the client sends a BDP ping at the right time.
-//  2. If, for example, the call to ss.S.Stop() happens after the RPC headers have been received at the
-//     server, then the TCP connection can shutdown gracefully when the server's socket closes. raceyness around checks if the error message when checking reading frames
+// possible conn closed errors.
 const possibleConnResetMsg = "connection reset by peer"
 const possibleEOFMsg = "error reading from server: EOF"
 
-func errMsgContainsPossibleStopMessages(err error) bool {
-	return strings.Contains(err.Error(), possibleConnResetMsg) || strings.Contains(err.Error(), possibleEOFMsg) || err == io.EOF
+// isConnClosedErr checks the error msg for possible conn closed messages. There
+// is a raceyness in the timing of when TCP packets are sent from client to
+// server, and when we tell the server to stop, so we need to check for both of
+// these possible error messages:
+//  1. If the call to ss.S.Stop() causes the server's sockets to close while
+//     there's still in-fight data from the client on the TCP connection, then
+//     the kernel can send an RST back to the client (also see
+//     https://stackoverflow.com/questions/33053507/econnreset-in-send-linux-c).
+//     Note that while this condition is expected to be rare due to the
+//     rpcStartedOnServer synchronization, in theory it should be possible,
+//     e.g. if the client sends a BDP ping at the right time.
+//  2. If, for example, the call to ss.S.Stop() happens after the RPC headers
+//     have been received at the server, then the TCP connection can shutdown
+//     gracefully when the server's socket closes.
+func isConnClosedErr(err error) bool {
+	errContainsConnResetMsg := strings.Contains(err.Error(), possibleConnResetMsg)
+	errContainsEOFMsg := strings.Contains(err.Error(), possibleEOFMsg)
+
+	return errContainsConnResetMsg || errContainsEOFMsg || err == io.EOF
 }
 
 func (l *lazyConn) Write(b []byte) (int, error) {
@@ -1043,10 +1051,10 @@ func (s) TestDetailedConnectionCloseErrorPropagatesToRpcError(t *testing.T) {
 	// the RPC has been started on it.
 	<-rpcStartedOnServer
 	ss.S.Stop()
-	// The precise behavior of this test is subject to raceyness around the timing of when TCP packets
-	// are sent from client to server, and when we tell the server to stop, so we need to account for both
-	// of these possible error messages
-	if _, err := stream.Recv(); err == nil || !errMsgContainsPossibleStopMessages(err) {
+	// The precise behavior of this test is subject to raceyness around the timing
+	// of when TCP packets are sent from client to server, and when we tell the
+	// server to stop, so we need to account for both possible error messages.
+	if _, err := stream.Recv(); err == nil || !isConnClosedErr(err) {
 		t.Fatalf("%v.Recv() = _, %v, want _, rpc error containing substring: %q OR %q", stream, err, possibleConnResetMsg, possibleEOFMsg)
 	}
 	close(rpcDoneOnClient)
@@ -6748,31 +6756,37 @@ func (s) TestRPCWaitsForResolver(t *testing.T) {
 func (s) TestHTTPHeaderFrameErrorHandlingHTTPMode(t *testing.T) {
 	// Non-gRPC content-type fallback path.
 	for httpCode := range transport.HTTPStatusConvTab {
-		doHTTPHeaderTest(t, transport.HTTPStatusConvTab[int(httpCode)], []string{
+		if err := doHTTPHeaderTest(t, transport.HTTPStatusConvTab[int(httpCode)], []string{
 			":status", fmt.Sprintf("%d", httpCode),
 			"content-type", "text/html", // non-gRPC content type to switch to HTTP mode.
 			"grpc-status", "1", // Make up a gRPC status error
 			"grpc-status-details-bin", "???", // Make up a gRPC field parsing error
-		})
+		}); err != nil {
+			t.Error(err)
+		}
 	}
 
 	// Missing content-type fallback path.
 	for httpCode := range transport.HTTPStatusConvTab {
-		doHTTPHeaderTest(t, transport.HTTPStatusConvTab[int(httpCode)], []string{
+		if err := doHTTPHeaderTest(t, transport.HTTPStatusConvTab[int(httpCode)], []string{
 			":status", fmt.Sprintf("%d", httpCode),
 			// Omitting content type to switch to HTTP mode.
 			"grpc-status", "1", // Make up a gRPC status error
 			"grpc-status-details-bin", "???", // Make up a gRPC field parsing error
-		})
+		}); err != nil {
+			t.Error(err)
+		}
 	}
 
 	// Malformed HTTP status when fallback.
-	doHTTPHeaderTest(t, codes.Internal, []string{
+	if err := doHTTPHeaderTest(t, codes.Internal, []string{
 		":status", "abc",
 		// Omitting content type to switch to HTTP mode.
 		"grpc-status", "1", // Make up a gRPC status error
 		"grpc-status-details-bin", "???", // Make up a gRPC field parsing error
-	})
+	}); err != nil {
+		t.Error(err)
+	}
 }
 
 // Testing erroneous ResponseHeader or Trailers-only (delivered in the first HEADERS frame).
@@ -6818,7 +6832,9 @@ func (s) TestHTTPHeaderFrameErrorHandlingInitialHeader(t *testing.T) {
 			errCode: codes.Unavailable,
 		},
 	} {
-		doHTTPHeaderTest(t, test.errCode, test.header)
+		if err := doHTTPHeaderTest(t, test.errCode, test.header); err != nil {
+			t.Error(err)
+		}
 	}
 }
 
@@ -6871,7 +6887,9 @@ func (s) TestHTTPHeaderFrameErrorHandlingNormalTrailer(t *testing.T) {
 	}
 	for _, test := range tests {
 		t.Run(test.name, func(t *testing.T) {
-			doHTTPHeaderTest(t, test.errCode, test.responseHeader, test.trailer)
+			if err := doHTTPHeaderTest(t, test.errCode, test.responseHeader, test.trailer); err != nil {
+				t.Error(err)
+			}
 		})
 
 	}
@@ -6882,7 +6900,9 @@ func (s) TestHTTPHeaderFrameErrorHandlingMoreThanTwoHeaders(t *testing.T) {
 		":status", "200",
 		"content-type", "application/grpc",
 	}
-	doHTTPHeaderTest(t, codes.Internal, header, header, header)
+	if err := doHTTPHeaderTest(t, codes.Internal, header, header, header); err != nil {
+		t.Fatal(err)
+	}
 }
 
 type httpServerResponse struct {
@@ -6947,13 +6967,13 @@ func (s *httpServer) start(t *testing.T, lis net.Listener) {
 		}
 		writer.Flush() // necessary since client is expecting preface before declaring connection fully setup.
 		var sid uint32
-		// Loop until conn is closed and framer returns possible Stop error messages
+		// Loop until framer returns possible conn closed errors.
 		for requestNum := 0; ; requestNum = (requestNum + 1) % len(s.responses) {
 			// Read frames until a header is received.
 			for {
 				frame, err := framer.ReadFrame()
 				if err != nil {
-					if !errMsgContainsPossibleStopMessages(err) {
+					if !isConnClosedErr(err) {
 						t.Errorf("Error at server-side while reading frame. got: %q, want: rpc error containing substring %q OR %q", err, possibleConnResetMsg, possibleEOFMsg)
 					}
 					return
@@ -7011,11 +7031,10 @@ func (s *httpServer) start(t *testing.T, lis net.Listener) {
 	}()
 }
 
-func doHTTPHeaderTest(t *testing.T, errCode codes.Code, headerFields ...[]string) {
-	t.Helper()
+func doHTTPHeaderTest(t *testing.T, errCode codes.Code, headerFields ...[]string) error {
 	lis, err := net.Listen("tcp", "localhost:0")
 	if err != nil {
-		t.Fatalf("Failed to listen. Err: %v", err)
+		return fmt.Errorf("failed to listen. Err: %v", err)
 	}
 	defer lis.Close()
 	server := &httpServer{
@@ -7024,7 +7043,7 @@ func doHTTPHeaderTest(t *testing.T, errCode codes.Code, headerFields ...[]string
 	server.start(t, lis)
 	cc, err := grpc.Dial(lis.Addr().String(), grpc.WithTransportCredentials(insecure.NewCredentials()))
 	if err != nil {
-		t.Fatalf("failed to dial due to err: %v", err)
+		return fmt.Errorf("failed to dial due to err: %v", err)
 	}
 	defer cc.Close()
 	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
@@ -7032,11 +7051,12 @@ func doHTTPHeaderTest(t *testing.T, errCode codes.Code, headerFields ...[]string
 	client := testpb.NewTestServiceClient(cc)
 	stream, err := client.FullDuplexCall(ctx)
 	if err != nil {
-		t.Fatalf("error creating stream due to err: %v", err)
+		return fmt.Errorf("error creating stream due to err: %v", err)
 	}
 	if _, err := stream.Recv(); err == nil || status.Code(err) != errCode {
-		t.Fatalf("stream.Recv() = _, %v, want error code: %v", err, errCode)
+		return fmt.Errorf("stream.Recv() = _, %v, want error code: %v", err, errCode)
 	}
+	return nil
 }
 
 func (s) TestClientCancellationPropagatesUnary(t *testing.T) {

